File: android/feature/catalog/src/main/kotlin/HomeShelves.kt
package catalog

import data.ProgressPoint
import data.ResumePoint
import model.MediaSet
import model.Progress
import model.WatchSnapshot

/** How many plates a row holds before the rest is left to its own shelf. */
const val HOME_ROW_LIMIT = 6

/**
 * One row of the start page. [seeAll] is the shelf "See all" opens, or
 * `null` while the row has nowhere to send it yet — Continue, until the
 * kept-shelves phase gives it a tab of its own. [total] is how much the row
 * is a window onto, for the heading: the plates on screen cannot say it on
 * their own.
 */
data class HomeRow(val title: String, val seeAll: String?, val total: Int, val content: RowContent)

/**
 * What a row draws. A show or a course arriving on Latest is a card for the
 * whole of it; an episode offered under Continue or Next up is not — it is
 * one set with a caption about this viewer's own place in it. Two shapes
 * rather than bending [Entry.Film] over an episode that is not a film.
 */
sealed interface RowContent {
    data class Entries(val entries: List<Entry>) : RowContent
    data class Sets(val cards: List<SetCard>) : RowContent
}

/** One set on Continue or Next up: what to play, what to say under it, and its own mark. */
data class SetCard(val set: MediaSet, val caption: String, val progress: Float?, val watched: Boolean)

/**
 * The rows the start page shows, from the library and this viewer's own
 * watch state — Continue and Next up first, then the three Latest shelves,
 * in that order because Continue and Next up are unlikely to correspond to
 * what a viewer opened this page to reload.
 */
fun homeRowsOf(shelves: List<Shelf>, watch: WatchSnapshot, limit: Int = HOME_ROW_LIMIT): List<HomeRow> {
    val collections = shelves.asSequence().flatMap { it.entries }.filterIsInstance<Entry.Collection>().toList()
    val underway = underwayOf(collections, indexById(shelves), watch, limit)
    val positions = watch.progress.associateBy { it.setId }
    val watchedIds = watch.watched.mapTo(HashSet()) { it.setId }

    val rows = mutableListOf<HomeRow>()

    if (underway.continues.isNotEmpty()) {
        rows += HomeRow(
            title = "Continue",
            seeAll = null,
            total = underway.continuesTotal,
            content = RowContent.Sets(
                underway.continues.map { set -> setCard(set, resumeLine(positions[set.setId]), positions, watchedIds) },
            ),
        )
    }

    if (underway.nextUp.isNotEmpty()) {
        rows += HomeRow(
            title = "Next up",
            seeAll = "Series",
            total = underway.nextUpTotal,
            content = RowContent.Sets(
                underway.nextUp.map { entry ->
                    // The captions differ within the row on purpose: one card
                    // is where the viewer stopped, the next is what follows
                    // an episode they finished.
                    val caption = if (entry.resume) resumeLine(positions[entry.set.setId]) else "Next up"
                    setCard(entry.set, caption, positions, watchedIds)
                },
            ),
        )
    }

    for (shelf in shelves) {
        rows += HomeRow(
            title = latestTitleFor(shelf.title),
            seeAll = shelf.title,
            total = shelf.entries.size,
            content = RowContent.Entries(newestFirst(shelf.entries, limit)),
        )
    }

    return rows
}

private fun setCard(
    set: MediaSet,
    caption: String,
    positions: Map<String, Progress>,
    watchedIds: Set<String>,
): SetCard {
    val progress = ResumePoint.watchedFraction(positions[set.setId]?.let { ProgressPoint(it.at, it.duration) })
    return SetCard(set, caption, progress?.toFloat(), set.setId in watchedIds)
}

/** Every set anywhere in [shelves], keyed by id — a film as much as an episode or a lesson. */
private fun indexById(shelves: List<Shelf>): Map<String, MediaSet> {
    val byId = HashMap<String, MediaSet>()
    for (shelf in shelves) {
        for (entry in shelf.entries) {
            when (entry) {
                is Entry.Film -> byId[entry.set.setId] = entry.set
                is Entry.Collection -> for (division in entry.divisions.asSequence().flatMap { it.walk() }) {
                    for (set in division.items) byId[set.setId] = set
                }
            }
        }
    }
    return byId
}

/**
 * A row's name, from the shelf it draws on.
 *
 * The shelf titles are the vocabulary a viewer already reads in the
 * masthead, so the rows borrow them rather than introducing a second set of
 * words for the same three things.
 */
private fun latestTitleFor(shelf: String): String = "Latest $shelf".lowercase()
    .replaceFirstChar(Char::uppercase)

/** Newest first, on a copy — the shelf keeps the order it was built in. */
private fun newestFirst(entries: List<Entry>, limit: Int): List<Entry> =
    entries.sortedByDescending(::arrivedAt).take(limit)

/**
 * When an entry last gained something.
 *
 * A collection is dated by its newest member, not its first: a series still
 * being uploaded keeps its place on the row, and one finished two years ago
 * does not hold the top of it for having been started recently.
 */
private fun arrivedAt(entry: Entry): Long = when (entry) {
    is Entry.Film -> entry.set.addedAt
    is Entry.Collection -> entry.divisions
        .asSequence()
        .flatMap { it.walk() }
        .flatMap { it.items.asSequence() }
        .maxOfOrNull(MediaSet::addedAt)
        ?: 0
}

File: web/public/lib/catalog/home-shelves.js
/**
 * What the start page shows, decided before anything is drawn.
 *
 * A library of six hundred titles has two facts about it worth landing on:
 * what arrived recently, and what was already underway. Everything here
 * derives those two from the catalog and the viewer's own state, and nothing
 * here touches the DOM — the rules are the part that can be got wrong, so
 * they are the part that is tested.
 *
 * The rule with any real substance is what a show's *next* episode is, and
 * it is written out in `nextInCollection`.
 */

import { flattenCollection } from "../library.js";
import { resumeAt } from "../resume-point.js";

/** How many cards a row holds before the rest is left to its own shelf:
 *  eight, the single row of posters the home page sets across its width. */
export const SHELF_LIMIT = 8;

/**
 * The five rows, from the library and what this viewer has watched.
 *
 * `progress` is the store's rows, newest first; `watchedAt` answers when a
 * set was finished, or `null`. They arrive as arguments rather than as an
 * import so this module can be tested without a store, and so the page
 * cannot accidentally read a third source of truth.
 *
 * @param {import("./home-shelves.js").HomeShelvesInput} from
 */
export function homeShelves({ library, byId, progress = [], watchedAt = () => null, limit = SHELF_LIMIT }) {
  const positions = new Map(progress.map((row) => [row.setId, row]));

  const underway = [];
  for (const collection of [...library.series, ...library.tutorials]) {
    const entry = nextInCollection(collection, positions, watchedAt);
    if (entry) underway.push(entry);
  }
  // The show touched most recently is the one most likely to be why this
  // page was opened.
  underway.sort((a, b) => b.touchedAt - a.touchedAt);
  const nextUp = underway.slice(0, limit);

  // One title is one card on this page. A show being watched would otherwise
  // appear twice within a single screen — as the episode in progress, and
  // again as the same episode under Next up.
  const shown = new Set(nextUp.map((entry) => entry.set.setId));
  const started = progress
    .filter((row) => resumeAt(row) !== null)
    .map((row) => byId.get(row.setId))
    .filter(Boolean);
  const underwaySets = started.filter((set) => !shown.has(set.setId));

  return {
    continues: underwaySets.slice(0, limit),
    nextUp,
    latestMovies: byArrival(library.movies, addedAt).slice(0, limit),
    latestSeries: byArrival(library.series, newestIn).slice(0, limit),
    latestCourses: byArrival(library.tutorials, newestIn).slice(0, limit),
    // How much is behind each row, for its heading. A row shows six; the
    // number is the whole of what "See all" would open, which is the thing
    // the six cannot tell a viewer on their own.
    // Continue counts every started title, as its own shelf does, including
    // the ones this page moved to Next up: the figure has to agree with the
    // page "See all" opens.
    totals: {
      continues: started.length,
      nextUp: underway.length,
      latestMovies: library.movies.length,
      latestSeries: library.series.length,
      latestCourses: library.tutorials.length,
    },
  };
}

/**
 * What to offer from one show or course, or `null` when it offers nothing.
 *
 * Four cases, in this order:
 *
 * 1. **Nothing watched** — not underway, so not on this row. It belongs to
 *    Latest, or to the shelf.
 * 2. **Something has a resume point** — the most recently touched of those.
 *    Being part-way through an episode beats whatever follows a finished
 *    one: the viewer is in the middle of *that* episode.
 * 3. **Everything watched is finished** — walk forward from the one finished
 *    most recently and take the first episode not already watched. Walking
 *    forward rather than searching from the start is what stops an episode
 *    seen out of order from pulling the show backwards.
 * 4. **Nothing follows** — the show is finished and leaves the row.
 *
 * A glance is not being underway. `resumeAt` already refuses a position in
 * the first half-minute, and a show whose only mark is one of those is
 * treated as untouched rather than offered a card it cannot caption.
 */
function nextInCollection(collection, positions, watchedAt) {
  const ordered = flattenCollection(collection);

  let resumeSet = null;
  let resumeTouch = -Infinity;
  let finishedSet = null;
  let finishedTouch = -Infinity;

  for (const set of ordered) {
    const row = positions.get(set.setId);
    if (row && resumeAt(row) !== null && row.updatedAt > resumeTouch) {
      resumeTouch = row.updatedAt;
      resumeSet = set;
    }
    const finished = watchedAt(set.setId);
    if (finished !== null && finished > finishedTouch) {
      finishedTouch = finished;
      finishedSet = set;
    }
  }

  const touchedAt = Math.max(resumeTouch, finishedTouch);
  if (!Number.isFinite(touchedAt)) return null;

  if (resumeSet) return { set: resumeSet, collection, resume: true, touchedAt };

  // The collection is already in display order. Continue through this one
  // array instead of flattening and searching again for each watched episode.
  const finishedIndex = ordered.findIndex((set) => set.setId === finishedSet.setId);
  for (let index = finishedIndex + 1; index < ordered.length; index++) {
    const candidate = ordered[index];
    if (watchedAt(candidate.setId) === null) {
      return { set: candidate, collection, resume: false, touchedAt };
    }
  }
  return null;
}

/** When a set arrived. Absent in a payload from an older player. */
const addedAt = (set) => Number(set?.addedAt) || 0;

/**
 * When a collection last gained something.
 *
 * Its newest episode, not its first: a series still being uploaded keeps its
 * place on Latest, and one finished two years ago does not hold the top of
 * the row for having been started recently.
 */
function newestIn(collection) {
  let newest = 0;
  for (const set of flattenCollection(collection)) {
    const at = addedAt(set);
    if (at > newest) newest = at;
  }
  return newest;
}

/** Newest first, on a copy — the caller's shelves keep their own order. */
function byArrival(entries, when) {
  return [...entries].sort((a, b) => when(b) - when(a));
}
